### PR TITLE
Express matcher effects in terms of MTL typeclasses

### PR DESCRIPTION
**Pull request by ArturGajowy**
## Overview
<sup>_What this PR does, and why it's needed_</sup>

This is not to say that the whole rholang uses the effects via MTL. This is to say that what's left should be trivial to fix :)

### JIRA ticket:
<sup>_Create it if there isn't one already._</sup>

### Notes
<sup>_Optional. Add any notes on caveats, approaches you tried that didn't work, or anything else._</sup>

### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

### CODE REVIEW

**Comment by ArturGajowy** [comment]
bors try

**Comment by ArturGajowy** [comment]
bors try

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/matcher/package.scala [hunk 1]
```diff
@@ -24,37 +28,38 @@ package object matcher {
   type NonDetFreeMapWithCost[A] = StateT[StreamWithCost, FreeMap, A]
   type StreamWithCost[A]        = StreamT[ErroredOrCostA, A]
 
+  type _freeMap[F[_]] = MonadState[F, FreeMap]
```
Comment (lukasz-golebiewski) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/matcher/package.scala]: Why the `_` prefixes? Are you trying to express something which should be expressed with an access modifier?

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/matcher/package.scala [hunk 2]
```diff
@@ -24,37 +28,38 @@ package object matcher {
...
+  type _cost[F[_]]    = MonadState[F, Cost]
+  type _error[F[_]]   = MonadError[F, OutOfPhlogistonsError.type]
+  type _short[F[_]]   = MonadError[F, Unit] //arises from and corresponds to the OptionT/StreamT in the stack
+
+  def _freeMap[F[_]](implicit ev: _freeMap[F]) = ev
```
[the remark this answers is not in the document]
Comment (ArturGajowy) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/matcher/package.scala]: For the reader of those declarations, probably. It's though very convenient on the call site, and in fact brings the convenience you gain with a custom typeclass to the type alias without:
 - a) having to copy the typeclass and have back-and-forth conversions between the custom typeclass and the cats equivalent
 - b) extending the cats typeclass and getting into even more trouble

TL;DR: Just like you do `Sync[F].delay { ... }` you'd (probably) like to do `_freeMap[F].modify(...)` without worrying that `freeMap` is a an alias (vs a type with a companion object) and thus has no implicit summoner method.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/matcher/package.scala (hunk 1 above)
Comment (ArturGajowy) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/matcher/package.scala]: This convention is inspired by the [eff library](http://atnos-org.github.io/eff/org.atnos.site.Tutorial.html#write-an-interpreter-for-your-program) and I use it because it elegantly avoids the name conflict between the `Cost` typeclass and the `_cost` effect. I could go with suffixes, but it'd:
 1. be more verbose
 2. expose implementation detail that's likely to change (_cost is likely to be writer in a while, and _error to be ApplicativeError/ApplicativeThrow vs MonadError)

In my mind, this convention means: this is an effect-type alias. Will be used similarly to capabilities, but for more generic and probably low-level/implementation stuff.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/matcher/package.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (ArturGajowy) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/matcher/package.scala]: I can add a
```
// Implicit summoner methods, just like `Monad.apply` on `Monad`'s companion object.
```
comment if it helps :)

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/matcher/package.scala (hunk 1 above)
Comment (lukasz-golebiewski) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/matcher/package.scala]: Makes sense. Shouldn't it read `_freeMapState` then?

Comment (ArturGajowy) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/matcher/package.scala]: I'd like to avoid the suffix for the same reason `_error` is not `_monadError` :)

Comment (rabbitonweb) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/matcher/package.scala]: it's a bit unusual for the type to be lowercased, i was confused when I saw the usage

Comment (rabbitonweb) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/matcher/package.scala]: you'r saying this is convention?

Comment (ArturGajowy) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/matcher/package.scala]: See my reply to Łukasz. It's up to us to shape our conventions. I've seen it in the wild though, in the `eff` lib's docs. I guess I'll add my explanation to a comment near those type aliases :)
